### PR TITLE
Колодкин Григорий. Лабораторная работа №1: ClangAST. Вариант 1.

### PR DESCRIPTION
**Pull request by KolodkinGrigorii**
Данный плагин выводит информацию о пользовательском типе данных: является ли структура наследственной, имеет ли какие-либо поля и методы и, если имеет, выводит информацию о них(тип данных, уровень доступа, является ли метод виртуальным/чисто виртуальным/переопределенным)

### CODE REVIEW

File: clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            if (method->isVirtual() && !method->hasAttr<clang::OverrideAttr>()) specs.push_back("virtual");
+            if (method->isPureVirtual()) specs.push_back("pure");
+            if (method->hasAttr<clang::OverrideAttr>()) specs.push_back("override");
+
+            std::ostringstream oss;
+            oss << method->getReturnType().getAsString() << "(";
+            for (unsigned i = 0; i < method->getNumParams(); ++i) {
+                if (i > 0) oss << ", ";
+                oss << method->getParamDecl(i)->getType().getAsString();
+            }
```
Comment (aobolensk) [inline clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp]: Please, consider using `llvm::interleaveComma` instead of this loop

File: clang/test/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_test.cpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  double area() override {
+    return width * height;
+  }
+};
+
+
+
+
+
+
```
Comment (aobolensk) [inline clang/test/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_test.cpp]: Is this needed?

File: clang/test/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_test.cpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+// CHECK-NEXT: |_Methods
+// CHECK-NEXT: | |_ area (double()|public|override)
+
+// CHECK: Rectangle -> Shape
+// CHECK-NEXT: |_Fields
+// CHECK-NEXT: | |_ width (double|public)
+// CHECK-NEXT: | |_ height (double|public)
+// CHECK-NEXT: |_Methods
+// CHECK-NEXT: | |_ area (double()|public|override)
+
```
Comment (aobolensk) [inline clang/test/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_test.cpp]: Please, add more tests for empty/classes structs, multiple inheritance, nested classes/structs and so on

**Review by aobolensk — CHANGES_REQUESTED** [review]

**Review by m-ly4 — CHANGES_REQUESTED** [review]
Please also add printing of friend classes and functions.

File: clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    }
+}
+
+class LabPluginVisitor final : public clang::RecursiveASTVisitor<LabPluginVisitor> {
+public:
+    explicit LabPluginVisitor(clang::ASTContext *context) : m_context(context) {}
+    bool VisitCXXRecordDecl(clang::CXXRecordDecl *record) {
+    if (record->isCompleteDefinition()) {
+        llvm::outs() << record->getName();
+        if (record->isPolymorphic() && !record->bases().empty()) {
```
Comment (Kuznetsov-Artyom) [inline clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp]: Need the condition "record->isPolymorphic()"? Read the description of this method carefully

File: clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp [hunk 5]
```diff
@@ -0,0 +1,112 @@
+#include <string>
+#include <sstream>
+#include <vector>
```
Comment (Kuznetsov-Artyom) [inline clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp]:
```suggestion
#include <sstream>
```

File: clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace kolodkin_g_lab1 {
```
Comment (Kuznetsov-Artyom) [inline clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp]: In the context of a plugin, then it is better to use an anonymous namespace

File: clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+std::string changeToString(clang::AccessSpecifier access) {
```
Comment (Kuznetsov-Artyom) [inline clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp]: The function name is not informative

File: clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    if (access == clang::AS_public) {
+        return "public";
+    } else if (access == clang::AS_protected) {
+        return "protected";
+    } else if (access == clang::AS_private) {
+        return "private";
+    } else {
+        return "";
+    }
+}
```
Comment (Kuznetsov-Artyom) [inline clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp]: Use switch case

File: clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+class LabPluginVisitor final : public clang::RecursiveASTVisitor<LabPluginVisitor> {
+public:
+    explicit LabPluginVisitor(clang::ASTContext *context) : m_context(context) {}
+    bool VisitCXXRecordDecl(clang::CXXRecordDecl *record) {
+    if (record->isCompleteDefinition()) {
+        llvm::outs() << record->getName();
+        if (record->isPolymorphic() && !record->bases().empty()) {
+            llvm::outs() << " -> " << record->bases().begin()->getType()->getAsCXXRecordDecl()->getName();
```
Comment (Kuznetsov-Artyom) [inline clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp]: What about multiple inheritance?

File: clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+                              << "|" << kolodkin_g_lab1::changeToString(field->getAccess()) << ")\n";
+            }
+        }
+
+        llvm::outs() << "|_Methods\n";
+        bool hasMethods = false;
+        for (const auto *method : record->methods()) {
+            if (method->isImplicit()) continue;
+
+            hasMethods = true;
```
Comment (Kuznetsov-Artyom) [inline clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp]:
```cpp
if (record->methods().empty()) {
  // code
} else {
  // code
}

File: clang/test/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_test.cpp [hunk 11]
```diff
@@ -0,0 +1,48 @@
+// RUN: %clang_cc1 -load %llvmshlibdir/LabPlugin_KolodkinGrigorii_FIIT3_ClangAST%pluginext -plugin LabPlugin_KolodkinGrigorii_FIIT3_ClangAST -fsyntax-only %s 2>&1 | FileCheck %s
+
+// CHECK: Shape
+// CHECK-NEXT: |_Fields
+// CHECK-NEXT: | |_ (no fields)
+// CHECK-NEXT: |_Methods
+// CHECK-NEXT: | |_ area (double()|public|virtual|pure)
```
Comment (Kuznetsov-Artyom) [inline clang/test/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_test.cpp]:
```suggestion
// CHECK: Shape
// CHECK-NEXT: |_Fields
// CHECK-NEXT: | |_ (no fields)
// CHECK-NEXT: |_Methods
// CHECK-NEXT: | |_ area (double()|public|virtual|pure)
struct Shape {
  virtual double area() = 0;
};
```

**Review by Kuznetsov-Artyom — CHANGES_REQUESTED** [review]

File: clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    } else {
+        return "";
+    }
+}
+
+class LabPluginVisitor final : public clang::RecursiveASTVisitor<LabPluginVisitor> {
+public:
+    explicit LabPluginVisitor(clang::ASTContext *context) : m_context(context) {}
+    bool VisitCXXRecordDecl(clang::CXXRecordDecl *record) {
+    if (record->isCompleteDefinition()) {
```
Comment (m-ly4) [inline clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp]:
```suggestion
        if (!record->isCompleteDefinition())
            return true;
```

File: clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+            llvm::outs() << "| |_ (no methods)\n";
+        }
+
+        llvm::outs() << "\n";
+    }
+    return true;
+    }
+
+private:
+    clang::ASTContext *m_context;
```
Comment (m-ly4) [inline clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp]: This field is not used. Should be removed then.
```suggestion
```

File: clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+    switch (static_cast<int>(access)) {
+        case static_cast<int>(clang::AccessSpecifier::AS_public):
+            return "public";
+        case static_cast<int>(clang::AccessSpecifier::AS_protected):
+            return "protected";
+        case static_cast<int>(clang::AccessSpecifier::AS_private):
+            return "private";
+        default:
+            return "";
+    }
```
Comment (m-ly4) [inline clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp]: switch-case well works with enum [class] values without explicit casts.

File: clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+class LabPluginVisitor final : public clang::RecursiveASTVisitor<LabPluginVisitor> {
+public:
+    explicit LabPluginVisitor(clang::ASTContext *context) {}
+    bool VisitCXXRecordDecl(const clang::CXXRecordDecl *record) {
+    if (!record->isCompleteDefinition()) 
+	return true;
```
Comment (m-ly4) [inline clang/compiler-course/kolodkin_g_lab_1_var_1/kolodkin_g_lab1_plugin.cpp]: Please use only spaces and no tabs in source files.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Review by aobolensk — APPROVED** [review]

**Review by Kuznetsov-Artyom — APPROVED** [review]
lgtm

**Review by m-ly4 — APPROVED** [review]

**Review by m-ly4 — CHANGES_REQUESTED** [review]
Please fix tests

**Review by m-ly4 — APPROVED** [review]
